### ISSUE TITLE
Invalid IAP assertion (HTTP 403: Forbidden) RELOAD

### ISSUE THREAD

**Issue by duxbuse** [GoogleCloudPlatform/jit-access#78]
Howdy guys just got through getting all the terraform to set this up. Then after authentication I get a mostly blank page. It even doesn't tell me who I am signed in as.

<img width="499" alt="image" src="https://github.com/GoogleCloudPlatform/jit-access/assets/6869131/37faa1eb-ec71-4f68-af15-02951f38caff">

When I look at the chrome console I see
`Failed to load resource: the server responded with a status of 403 (Forbidden)  /api/policy:1`
Navigating to that page also shows:
```json
{"message":"Missing header: X-JITACCESS"}
```
But this might be because I am not supposed to navigate directly so might not be relevant.

At first I thought it might be an issue since the cloud identity api is not protected by VPC SC, so I have updated the dns entry to allow it to use the `private.googleapis.com` so I know its not a VPC SC thing.

Any ideas on how to resolve this is most appreciated.

**Comment by jpassing** [comment]
Did you deploy on Cloud Run or on AppEngine?

For security reasons, the error message doesn't contain any details about what's wrong with the IAP assertion. But the (AppEngine/Cloud Run) logs should contain a more verbose error message.  Could you have a look and share the error details?

As you mentioned VPC-SC, I suppose one possible explanation might be lack of outbound internet access: To verify the IAP assertion, the app needs to fetch the IAP public keys from `https://www.gstatic.com/iap/verify/public_key-jwk`... but that should show in the logs.

> But this might be because I am not supposed to navigate directly so might not be relevant.

Yes, you can ignore that error, it's expected if you try to hit the REST endpoints directly.

**Comment by duxbuse** [comment]
I deployed with cloud run.

Had a look at the logs, there are no errors but a couple of warnings:

```
The SMTP configuration is incomplete
```

and

```json
{
  "insertId": "646fff80000b99cf835512ff",
  "httpRequest": {
    "requestMethod": "GET",
    "requestUrl": "https://jitaccess.iag.com.au/api/policy",
    "requestSize": "2820",
    "status": 403,
    "responseSize": "705",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36",
    "remoteIp": "**xx.xx.xx.xx**",
    "serverIp": "**xx.xx.xx.xx**",
    "latency": "0.130215961s",
    "protocol": "HTTP/1.1"
  },
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
      "location": "australia-southeast1",
      "service_name": "jitaccess",
      "project_id": "xxx-xxx",
      "configuration_name": "jitaccess",
      "revision_name": "jitaccess-00002-9s8"
    }
  },
  "timestamp": "2023-05-26T00:38:24.421965Z",
  "severity": "WARNING",
  "labels": {
    "instanceId": "004d9db0be6e814c75bc469318d434d4dfe21fcbc1a517274a9b758aec4b4fd9c25928e9c0f13a2edd5bb0137e029ae726b6aa09f1f338f0b3f99a7715b34e59dd286e"
  },
  "logName": "projects/xxx-xxx/logs/run.googleapis.com%2Frequests",
  "trace": "projects/xxx-xxx/traces/a619b5648decc79f9c5df492c13f9993",
  "receiveTimestamp": "2023-05-26T00:38:24.979113232Z",
  "spanId": "17382944962019312285"
}
```

**Comment by duxbuse** [comment]
Actually I think it might be proxy related. 

We can only get to the internet via our corp proxy. So will need to figure out how to add a http proxy to IAP 🤔

**Comment by jpassing** [comment]
I haven't tried this in Cloud Run, but the application does support proxy servers. What you'd need to do is update the `app.yaml` (used to deploy the container) to [pass 3 extra command line arguments to the container](https://cloud.google.com/run/docs/configuring/containers#configure_container_entrypoint_command_and_arguments): `-DproxyHost=proxy.example.com -DproxyPort=3128 -DproxySet=true`.

JIT Access will then use this proxy server for all outbound connections, including API calls.

If your proxy does TLS inspection, then a bit more configuration is required.

**Comment by duxbuse** [comment]
yeah will need to config certs/tls

**Comment by jpassing** [comment]
Sorry for the delayed reply.

> But does that mean that it all runs from the cloud run and the IAP client doesnt make any requests?

In each request that IAP forwards to the Cloud Run app, it [includes a JWT in a HTTP request header](https://cloud.google.com/iap/docs/signed-headers-howto). JIT Access then takes this JWT and attempts to verify it -- but this seems to fail in your environment.

As you noted, JIT Access currently doesn't write any logs when JWT verification fails... that's not good, so I prepared a fix to change that (#80). It's now merged into the `master` branch. If you redeploy from the `master` branch, you should now see some more error details about why JWT verification fails in your environment.

**Comment by duxbuse** [comment]
Ok great thanks for that, I am now getting:
`Verifying IAP assertion failed: Expected audience does not match`

Not really sure what this means though

**Comment by jpassing** [comment]
The expected audience is [derived from the project number and backend ID](https://cloud.google.com/iap/docs/signed-headers-howto#verifying_the_jwt_payload). If it doesn't match, then it's probaby because JIT access uses an incorrect backend ID.

Could you check the value of the `IAP_BACKEND_SERVICE_ID` environment variable in `app.yaml`  and verify that if it matches the ID returned by `gcloud compute backend-services describe jitaccess-backend --global --format 'value(id)'`?

**Comment by duxbuse** [comment]
Ahh yes as part of my testing I did end up re creating the backend and forgot to update the ID, now its working better but when trying to list projects I get:
 `Listing available projects failed: Denied access to scope 'organizations/XXXXX': 403 Forbidden`
 But I have `roles/resourcemanager.organizationViewer`
 
 Perhaps I didn't quite understand how it works.
 
 We grant the JIT SA `roles/cloudasset.viewer` everywhere. But when I am trying to select a project to grant permissions in, that seems to be done with my own personal account which makes sense in that I can only list the projects I have permissions to see. So I granted myself org viewer (I already had ord admin but figured it wouldn't hurt) but was still getting the 403. Is this a https://cloud.google.com/vpc-service-controls/docs/supported-products#table_cloud_asset_api VPC SC issue where:
 
**Limitations** 

- VPC Service Controls doesn't support access to folder-level or organization-level Cloud Asset API resources from resources and clients inside a service perimeter.  VPC Service Controls protects project-level Cloud Asset API resources. You can specify an egress policy to prevent access to project-level Cloud Asset API resources from projects inside the perimeter.
- VPC Service Controls doesn't support adding folder-level or organization-level Cloud Asset API resources into a service perimeter. You cannot use a perimeter to protect folder-level or organization-level Cloud Asset API resources. To manage Cloud Asset Inventory permissions at the folder or organization level, we recommend using IAM.
- You cannot export assets at the folder or organization level into destinations inside a service perimeter.
- You cannot create real-time feeds for assets at the folder or organization level with a Pub/Sub topic inside a service perimeter.

So basically regardless of your perms because my cloud run is inside a VPC I can never `export assets at the folder or organization level`?
I guess is there a way to list the projects using the `resourcemanager.projects.get` rather than `cloudasset.assets.listResource`?

**Comment by jpassing** [comment]
> Is this a https://cloud.google.com/vpc-service-controls/docs/supported-products#table_cloud_asset_api VPC SC issue where

Yes, if the Cloud Run app is inside a perimeter, then that's probably the issue. 

> So basically regardless of your perms because my cloud run is inside a VPC I can never export assets at the folder or organization level?

Yes, and I realize that this limitation needs to be documented better. 

Using JIT Access to manage access for the same project might still work. But to let it manage access for an entire folder/organization, you'd need to deploy it outside the perimeter and [grant it ingress access to the perimeter](https://github.com/GoogleCloudPlatform/jit-access/wiki/Granting-access-to-VPC-service-perimeters). 

JIT Access used to run on App Engine only which made it impossible to run it inside a perimeter. Now that it also works on Cloud Run, you can deploy it inside a perimeter, but unfortunately that creates new issues...

> I guess is there a way to list the projects using the resourcemanager.projects.get rather than cloudasset.assets.listResource?

The Resource Manager APIs let you query which resources _you_ (=the caller) can access while the Asset API lets you query what _other users_ can access. From the perspective of the JIT Access app (which runs as a service account) that means: By using the Resource Manager APIs, the app would only be able to find out what it can access itself (which isn't very intersting), but it wouldn't be able to find out what the user can access.

**Comment by duxbuse** [comment]
Actually it just ended up being that the `XXXX-compute@developer.gserviceaccount.com` needed `roles/cloudasset.viewer`

Even though the caller is `service-XXXX@serverless-robot-prod.iam.gserviceaccount.com` it impersonates the node SA I guess?

Perhaps I haven't set up cloud run to use the right SA, ill have a play with it

**Comment by duxbuse** [comment]
Thanks for you help @jpassing got it all working now super amazing.

I had to write a fair bit of Terraform to get it all working. Would you be interested in the terraform to deploy this or are you happy with the gcloud commands only. Its not always a 1-1 translation
